### ISSUE TITLE
Using trtexec to generate engine file

### ISSUE THREAD

**Issue by LieceC** [NVIDIA/TensorRT#1929]
Hello, I can succesfully generate an int8 engine file of my pre-trained model using the trtexec command through an onnx representation. I can also run it succesfully on the given dimensions using the **python bindings** of TensorRT. However, once I try to compute the inference using the TensorRT c++ API directly. I face multiples errors :

`3: [executionContext.cpp::execute::475] Error Code 3: API Usage Error (Parameter check failed at: runtime/api/executionContext.cpp::execute::475, condition: batchSize > 0 && batchSize <= mEngine.getMaxBatchSize()`

One of the problem is that it seems to set the maxBatchSize to 1. Effectively, the onnx model is created with a batch of 2 and we can't set the variable to 2 throught the trtExec command so I'm not completely how to fix this part.

The second problem is that the detection models as multiple outputs, I'm not sure how to verify the output size on the inference accurately and if I should stock all output during inference ? Only the last one is interesting for me.

Here is a snippet of the function I use to run the inference : 

`bool Engine::runInference(const std::vector<cv::Mat> &inputFaceChips, std::vector<std::vector<float>>& featureVectors) {
    auto dims = m_engine->getBindingDimensions(0);
    std::cout<<m_engine->getBindingDimensions(4)<<std::endl; //detection layer output
    Dims3 outputL = {static_cast<int32_t>(inputFaceChips.size()),m_engine->getBindingDimensions(4).d[1], m_engine->getBindingDimensions(4).d[2]};
    std::cout<<outputL<<std::endl;
    Dims4 inputDims = {static_cast<int32_t>(inputFaceChips.size()), dims.d[1], dims.d[2], dims.d[3]};
    m_context->setBindingDimensions(0, inputDims);

    if (!m_context->allInputDimensionsSpecified()) {
        throw std::runtime_error("Error, not all input dimensions specified.");
    }

    auto batchSize = static_cast<int32_t>(inputFaceChips.size());
    // Only reallocate buffers if the batch size has changed
    if (m_prevBatchSize != inputFaceChips.size()) {

        m_inputBuff.hostBuffer.resize(inputDims);
        m_inputBuff.deviceBuffer.resize(inputDims);

        m_outputBuff.hostBuffer.resize(outputL);
        m_outputBuff.deviceBuffer.resize(outputL);

        m_prevBatchSize = batchSize;
    }

    auto* hostDataBuffer = static_cast<float*>(m_inputBuff.hostBuffer.data());

    for (size_t batch = 0; batch < inputFaceChips.size(); ++batch) {
        auto image = inputFaceChips[batch];

        // Preprocess code
        image.convertTo(image, CV_32FC3, 1.f / 255.f);
        cv::subtract(image, cv::Scalar(0.5f, 0.5f, 0.5f), image, cv::noArray(), -1);
        cv::divide(image, cv::Scalar(0.5f, 0.5f, 0.5f), image, 1, -1);

        // NHWC to NCHW conversion
        // NHWC: For each pixel, its 3 colors are stored together in RGB order.
        // For a 3 channel image, say RGB, pixels of the R channel are stored first, then the G channel and finally the B channel.
        // https://user-images.githubusercontent.com/20233731/85104458-3928a100-b23b-11ea-9e7e-95da726fef92.png
        int offset = dims.d[1] * dims.d[2] * dims.d[3] * batch;
        int r = 0 , g = 0, b = 0;
        for (int i = 0; i < dims.d[1] * dims.d[2] * dims.d[3]; ++i) {
            if (i % 3 == 0) {
                hostDataBuffer[offset + r++] = *(reinterpret_cast<float*>(image.data) + i);
            } else if (i % 3 == 1) {
                hostDataBuffer[offset + g++ + dims.d[2]*dims.d[3]] = *(reinterpret_cast<float*>(image.data) + i);
            } else {
                hostDataBuffer[offset + b++ + dims.d[2]*dims.d[3]*2] = *(reinterpret_cast<float*>(image.data) + i);
            }
        }
    }

    // Copy from CPU to GPU
    auto ret = cudaMemcpyAsync(m_inputBuff.deviceBuffer.data(), m_inputBuff.hostBuffer.data(), m_inputBuff.hostBuffer.nbBytes(), cudaMemcpyHostToDevice, m_cudaStream);
    if (ret != 0) {
        return false;
    }

    std::vector<void*> predicitonBindings = {m_inputBuff.deviceBuffer.data(), m_outputBuff.deviceBuffer.data()};

    // Run inference.
    bool status = m_context->execute(2,predicitonBindings.data());
    if (!status) {
        return false;
    }

    // Copy the results back to CPU memory
    ret = cudaMemcpyAsync(m_outputBuff.hostBuffer.data(), m_outputBuff.deviceBuffer.data(), m_outputBuff.deviceBuffer.nbBytes(), cudaMemcpyDeviceToHost, m_cudaStream);
    if (ret != 0) {
        std::cout << "Unable to copy buffer from GPU back to CPU" << std::endl;
        return false;
    }

    ret = cudaStreamSynchronize(m_cudaStream);
    if (ret != 0) {
        std::cout << "Unable to synchronize cuda stream" << std::endl;
        return false;
    }

    return true;
}`

I use cuda 10.2.00 and TensorRT 8.2.
I use a slightly modified version of the code present at : https://github.com/cyrusbehr/tensorrt-cpp-api to test my inference.

**Comment by nvpohanh** [comment]
Please use `executeV2()` instead of `execute()`. `execute()` has been deprecated.

**Comment by LieceC** [comment]
Hello, I initially used executev2 but I had an issue with empty bindings
and couldn't understand where that come from

**Comment by nvpohanh** [comment]
Alternatively, you can call `execute()` with `batchSize` field set always to 1 because trtexec builds the engine using explicit-batch-dim mode, so you should use setBindingDimensions() to set the input shapes instead of using the `batchSize` field. Thanks

**Comment by LieceC** [comment]
> Alternatively, you can call `execute()` with `batchSize` field set always to 1 because trtexec builds the engine using explicit-batch-dim mode, so you should use setBindingDimensions() to set the input shapes instead of using the `batchSize` field. Thanks

Hello thank you for the hints, I tried these methods, and this is the outputs after printing the first and last binding dimensions: 
`m_context->setBindingDimensions(0, inputDims);
std::cout<<m_engine->getBindingDimensions(4)<<std::endl;
std::cout<<m_engine->getBindingDimensions(0)<<std::endl;`

`(2, 1165248, 21)
(2, 3, 4352, 4352)`

So far the error using execute stay the same : `
3: [executionContext.cpp::execute::475] Error Code 3: API Usage Error (Parameter check failed at: runtime/api/executionContext.cpp::execute::475, condition: batchSize > 0 && batchSize <= mEngine.getMaxBatchSize()
)`

Using executeV2 however, the errors is different : `[executionContext.cpp::executeInternal::652] Error Code 1: Cuda Runtime (misaligned address)`. Might it comes from the numbers of bindings dimensions ? I actually got 4 output dimensions (3 from backbone 1 from detector) but only 1 is interesting in my case and I only allocated for this one.

Thank you

**Comment by nvpohanh** [comment]
> I actually got 4 output dimensions (3 from backbone 1 from detector) but only 1 is interesting in my case and I only allocated for this one.

I think you will have to allocate the buffers for all the input and output bindings, even if you don't need them.

**Comment by LieceC** [comment]
I see I'll try this, thank you!

**Comment by LieceC** [comment]
The inference is now working correctly :+1:

**Comment by nvpohanh** [comment]
Glad to hear that it worked! We are also modifying the error message in enqueue() to make it more understandable that enqueueV2() should be used. The change will be part of next TRT release. :)
